utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2112 @@
+    // else process the file if its processing epoch matches
+    // the current epoch. Typically the #of epochs should be
+    // around 5-10
+    if (!(bfile->gc_epoch_ == -1 ||
+          (uint64_t)bfile->gc_epoch_ == current_epoch_)) {
+      continue;
+    }
+
+    files_processed++;
+    // reset the epoch
+    bfile->gc_epoch_ = current_epoch_ + next_epoch_increment;
+    to_process.push_back(bfile);
+    if (files_processed >= files_to_collect) break;
+  }
+
+  // in this collect the set of files, which became obsolete
+  std::vector<std::shared_ptr<BlobFile>> obsoletes;
+  for (auto bfile : to_process) {
+    // File can be obsolete. File can be Open for writes, or closed
+    std::string pn = bfile->PathName();
+
+    {
+      ReadLock rl_f(&bfile->mutex_);
+      // in a previous pass, this file was marked obsolete
+      // or this file is still active for appends.
+      if (bfile->Obsolete() || !bfile->Immutable()) continue;
+
+      std::string reason;
+      bool shouldgc = shouldGCFile_locked(bfile, tt, last_id, &reason);
+      if (!shouldgc) {
+        Log(InfoLogLevel::DEBUG_LEVEL, db_options_.info_log,
+            "File has been skipped for GC ttl %s %d %d reason='%s'", pn.c_str(),
+            tt, bfile->GetTTLRange().second, reason.c_str());
+        continue;
+      }
+
+      Log(InfoLogLevel::INFO_LEVEL, db_options_.info_log,
+          "File has been chosen for GC ttl %s %d %d reason='%s'", pn.c_str(),
+          tt, bfile->GetTTLRange().second, reason.c_str());
+    }
+
+    Status s = writeBatchOfDeleteKeys(bfile, tt);
+    if (!s.ok()) continue;
+
+    bfile->setCanBeDeleted();
+    obsoletes.push_back(bfile);
+  }
+
+  if (!obsoletes.empty()) {
+    WriteLock wl(&mutex_);
+    for (auto bfile : obsoletes) {
+      obsolete_files_.push_front(bfile);
+      // remove from global list so writers
+      blob_files_.erase(bfile->BlobFileNumber());
+    }
+  }
+
+  // reschedule
+  return std::make_pair(true, -1);
+}
+
+BlobFile::BlobFile()
+    : parent_(nullptr),
+      file_number_(0),
+      blob_count_(0),
+      gc_epoch_(-1),
+      file_size_(0),
+      deleted_count_(0),
+      deleted_size_(0),
+      closed_(false),
+      can_be_deleted_(false),
+      ttl_range_(std::make_pair(0, 0)),
+      time_range_(std::make_pair(0, 0)),
+      sn_range_(std::make_pair(0, 0)),
+      last_access_(-1),
+      last_fsync_(0),
+      header_valid_(false) {}
+
+BlobFile::BlobFile(const BlobDBImpl* p, const std::string& bdir, uint64_t fn)
+    : parent_(p),
+      path_to_dir_(bdir),
+      file_number_(fn),
+      blob_count_(0),
+      gc_epoch_(-1),
+      file_size_(0),
+      deleted_count_(0),
+      deleted_size_(0),
+      closed_(false),
+      can_be_deleted_(false),
+      ttl_range_(std::make_pair(0, 0)),
+      time_range_(std::make_pair(0, 0)),
+      sn_range_(std::make_pair(0, 0)),
+      last_access_(-1),
+      last_fsync_(0),
+      header_valid_(false) {}
+
+BlobFile::~
```

Comment:
this is a very standard pattern used in many software. When you call a function
after holding the mutex, you name that function _locked. When the function takes
the lock, you have the simpler name. It helps in reasoning about multi-threaded code
a lot. I have changed it to Locked()